src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs

```diff
@@ -1324,15 +1316,40 @@ .. chosenResolvedItem.Suppressions
                         continue;
                     }
 
+                    // See if a dependency with the same version and suppressions has already been chosen
+                    if (resolvedDependencyGraphItems.TryGetValue(childLibraryDependencyIndex, out ResolvedDependencyGraphItem? childResolvedDependencyGraphItem)
+                        && childResolvedDependencyGraphItem.LibraryRangeIndex == childLibraryRangeIndex
+                        && childResolvedDependencyGraphItem.Suppressions.Count == 1
+                        && childResolvedDependencyGraphItem.Suppressions[0].Count == 0
+                        && HasCommonAncestor(childResolvedDependencyGraphItem.Path, currentDependencyGraphItem.Path))
+                    {
+                        childResolvedDependencyGraphItem.Parents ??= new HashSet<LibraryRangeIndex>();
+
+                        if (!childResolvedDependencyGraphItem.IsRootPackageReference)
+                        {
+                            // Keep track of the parents of this item
+                            childResolvedDependencyGraphItem.Parents?.Add(currentDependencyGraphItem.LibraryRangeIndex);
```

Comment:
I don't believe it's possible for Parents to be null here since we instantiate it on line 1326 if it is null. So, we should remove the operator to make that clear.

```suggestion
                            childResolvedDependencyGraphItem.Parents.Add(currentDependencyGraphItem.LibraryRangeIndex);
```

Reply:
Good idea, I've cleaned up the Parents stuff, let me know if you like it better now.

Reply:
The ?. operator is still present when dereferencing the Parents which implies that Parents can be null and looking at the changes to the dependency graph item, it looks like there is a case where it can be null. In that case, we'll skip adding the currentDependencyGraphItem.LibraryRangeIndex to the Parents of the childResolvedDependencyGraphItem. This implicit relationship between whether the HashSet is instantiated and whether parents are kept track of worries me for code maintenance and clarity. Is it intentional that we would have a non-root package reference that has null parents that we don't want to keep track of even if the currentDependencyGraphItem has a LibraryRangeIndex? Could this relationship be represented more clearly through tests or could we add a comment? Alternatively, would it make sense to have an empty hash set for the Parents to avoid special casing when it's null or would that break things?

Reply:
@jeffkl @jebriede please follow up offline if necessary. This PR is needed to fix a high priority perf regression